lms/services/oauth_http.py

```diff
@@ -82,19 +92,38 @@ def get_access_token(self, token_url, redirect_uri, auth, authorization_code):
             },
         )
 
-    def refresh_access_token(self, token_url, redirect_uri, auth):
+    def refresh_access_token(
+        self, token_url, redirect_uri, auth, prevent_concurrent_refreshes=False
+    ):
         """
         Make a refresh token request and save the new token in the DB.
 
         Send an OAuth 2.0 "refresh token request"
         (https://datatracker.ietf.org/doc/html/rfc6749#section-6) to get a new
         access token for the current user and save it to the DB.
 
+        :raise ConcurrentTokenRefreshError: if this token is being refreshed concurrently by another request
         :raise OAuth2TokenError: if we don't have a refresh token for the user
         :raise ExternalRequestError: if the HTTP request fails
         :raise ValidationError: if the server's access token response is invalid
         """
-        refresh_token = self._oauth2_token_service.get(self.service).refresh_token
+        old_token = self._oauth2_token_service.get(self.service)
+
+        # If the "old" token is already current, just return immediately.
+        if (
+            old_token.access_token
+            and datetime.utcnow() - old_token.received_at < timedelta(seconds=30)
+        ):
+            return old_token.access_token
+
+        if prevent_concurrent_refreshes:
+            # Prevent concurrent refresh attempts. If the client gets this, it
+            # should wait briefly and try again, at which point it should find the
+            # refreshed token already available and skip the refresh.
```

Comment:
Maybe move this comment into the `except` block, before the `raise` statement, then it might make more sense what "this" refers to in "If the client gets this..."